### PR TITLE
rebase: Throw OutOfMemoryError when gmp tries to allocate memory beyond the limit

### LINKED ISSUE #8286: ^(big(foo), bar) can cause SIGBART
Example:

```
julia>^(big(96608869069402268615522366320733234710),16374500563449903721)
gmp: overflow in mpz type

signal (6): Aborted
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 1638697868)
__gmpz_n_pow_ui at /home/dluu/dev/julia/usr/bin/../lib/libgmp.so (unknown line)
^ at gmp.jl:337
bigint_pow at gmp.jl:346
^ at gmp.jl:351
jlcall_^;60129 at  (unknown line)
jl_apply at /home/dluu/dev/julia/src/julia.h:988
jl_trampoline at /home/dluu/dev/julia/src/builtins.c:821
jl_apply at /home/dluu/dev/julia/src/julia.h:988
jl_apply_generic at /home/dluu/dev/julia/src/gf.c:1592
unknown function (ip: 2066563356)
unknown function (ip: 2066564620)
unknown function (ip: 2066567534)
unknown function (ip: 2066563698)
unknown function (ip: 2066658267)
jl_toplevel_eval at /home/dluu/dev/julia/usr/bin/../lib/libjulia-debug.so (unknown line)
jl_f_top_eval at /home/dluu/dev/julia/src/builtins.c:425
eval_user_input at REPL.jl:54
jlcall_eval_user_input;60109 at  (unknown line)
jl_apply at /home/dluu/dev/julia/src/julia.h:988
jl_trampoline at /home/dluu/dev/julia/src/builtins.c:821
jl_apply at /home/dluu/dev/julia/src/julia.h:988
jl_apply_generic at /home/dluu/dev/julia/src/gf.c:1592
anonymous at task.jl:96
jl_apply at /home/dluu/dev/julia/src/julia.h:988
jl_trampoline at /home/dluu/dev/julia/src/builtins.c:821
unknown function (ip: 2066609306)
unknown function (ip: 2066612155)
unknown function (ip: 2066610812)
unknown function (ip: 2066610910)
jl_handle_stack_switch at /home/dluu/dev/julia/usr/bin/../lib/libjulia-debug.so (unknown line)
julia_trampoline at /home/dluu/dev/julia/usr/bin/../lib/libjulia-debug.so (unknown line)
unknown function (ip: 4203809)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 4199369)
Aborted (core dumped)
```

This is probably not specific to `^`, since this happens when gmp tries to `rellaoc` too much space:

```
  if (sizeof (mp_size_t) == sizeof (int))
    {
      if (UNLIKELY (new_alloc > ULONG_MAX / GMP_NUMB_BITS))
        {
          fprintf (stderr, "gmp: overflow in mpz type\n");
          abort ();
        }
    }
  else
    {
      if (UNLIKELY (new_alloc > INT_MAX))
        {
          fprintf (stderr, "gmp: overflow in mpz type\n");
          abort ();
        }
    }
```

It seems a bit impolite to crash julia because of this. What do people think is a good fix?
1. Before calling gmp, try to figure out of the calculation will result in trying to `realloc` to much space.
2. Catch `SIGABRT` and throw a julia error
3. Patch gmp to return a null pointer instead of throwing `SIGABRT`
4. ???

I don't like `1` because it requires doing something different for each function (I think?), but `2` and `3` have their own downsides.

### PR DESCRIPTION
**Pull request by musm**
rebased https://github.com/JuliaLang/julia/pull/29029 
and closes  #8286
courtesy @grep0

### CODE REVIEW

**Comment by musm** [comment]
Actually is anyone able to trigger the  original bug?

File: base/gmp.jl [hunk 1]
```diff
@@ -81,6 +81,14 @@ julia> big"313"
 """
 BigInt(x)
 
+"""
+    HAS_ALLOC_OVERFLOW_FUNCTION
+
+A reference that holds a boolean, if true, indicating julia is linked with a patched GMP that
+does not abort on huge allocation and throws OutOfMemoryError instead.
+"""
+const HAS_ALLOC_OVERFLOW_FUNCTION = Ref(false)
```
Comment (musm) [inline base/gmp.jl]: Is this name fine? Uppercase and all?

**Comment by musm** [comment]
I'm building julia but for some reason it's not picking up patches to GMP. My Make.user file looks like
```XC_HOST=x86_64-w64-mingw32
MARCH=native
USE_BINARYBUILDER_LLVM=1
USE_BINARYBUILDER_OPENBLAS=1
USE_BINARYBUILDER_SUITESPARSE=1```

Do  I have to do anything special to make sure julia builds the patched GMP file?

**Comment by musm** [comment]
Alright this PR is ready to be merged when ready. I confirmed also that it works locally.

**Comment by KristofferC** [comment]
Restarted it

**Comment by musm** [comment]
thanks all green now

**Comment by musm** [comment]
@JeffBezanson you approved the original PR, does this now look good to merge?

**Comment by doronbehar** [comment]
Hey all and sorry for bumping an old thread. I'm trying to collect what libraries that Julia depends upon, haven't included the patches you suggested them. I noticed that GMP 6.2.0, which was released long enough after this PR, hasn't included the patch here. See:

https://gmplib.org/repo/gmp-6.2/file/tip/memory.c#l39

Is my observation correct? I'm curios because I'm trying to package Julia for my distro and I want to know what libraries it should be safe enough to use our distro's versions vs your patched versions.

**Comment by nalimilan** [comment]
Indeed https://github.com/JuliaLang/julia/issues/8286 was marked as `upstream`, but nothing happened AFAICT.

**Comment by doronbehar** [comment]
Is there a link to an upstream patch submission? No feedback at all from their side? I even consider nagging them myself (yea I know it's not easy since it's a mailing list and not a GitHub PR) but I need something to hold on, not to mention I'd need to check the patch applies to their current version...

**Comment by nalimilan** [comment]
I mean that AFAICT nobody reported it upstream so far. Filing an issue with our patch it probably better than doing nothing, even if the patch may be slightly outdated.

**Comment by StefanKarpinski** [comment]
Agree. If you wouldn't mind doing that, it would be greatly appreciated.

**Comment by cmcaine** [comment]
I emailed this to the gmp-bugs mailing list today because it didn't look like anyone else had done it yet.

If the mailing list software didn't hate gmail then it should be publicly visible in the October archives in a few days: https://gmplib.org/list-archives/gmp-bugs/2020-October/thread.html

**Comment by musm** [comment]
@cmcaine  I opened an upstream request last month actually https://gmplib.org/list-archives/gmp-bugs/2020-September/004865.html   looks like  there is a library bug, but the resolution wasn't clear on which route the library developers are planning on going regarding this patch.

**Comment by cmcaine** [comment]
Woops! I don't know how I missed that :(

**Comment by StefanKarpinski** [comment]
Not much you can do: chances are if you get to the point of OOMing, the process is going to crash, we just want the ability to do it more gracefully than a C abort call, which is something that a library should really never do. If GMP wants to be super careful, it could free anything that was allocated before returning an error, but it would also be fine to document that when an OOM error is returned some allocated memory may be lost.

**Comment by nalimilan** [comment]
Just in case somebody qualified wants to have a look: this hasn't been upstreamed yet despite discussions on the mailing list.

**Pull request merged**